### PR TITLE
[py] Do not close externally provided log_output streams

### LINKED ISSUE #15629: [🐛 Bug]: [py] Service logging to console not working correctly
### Description

(sorry for the vague title. I'll rename it once I figure out what is happening)

Something weird is going on when I enable logging to the console in the Service class.

The following code should work properly. It is creating a driver instance with logging to `sys.stdout` enabled in the `Service` class. After instantiating the driver, it quits, and then creates another driver instance. This 2nd driver instance is totally unrelated to the first one, except for the fact that they both log to `sys.stdout`. This should work because `driver.quit()` closes the `sys.stdout` file descriptor, so there should be no issue re-using it in the next instance:

```
import sys
from selenium import webdriver
from selenium.webdriver.chrome.service import Service

service = Service(log_output=sys.stdout)
driver = webdriver.Chrome(service=service)
driver.quit()
service2 = Service(log_output=sys.stdout)
driver2 = webdriver.Chrome(service=service2)
```

When I run this, I get:

```
Starting ChromeDriver 135.0.7049.84 (6c019e56001911b3fd467e03bf68c435924d62f4-refs/branch-heads/7049@{#1778}) on port 52771
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully on port 52771.
Traceback (most recent call last):
  File "/home/cgoldberg617/code/selenium/py/sel.py", line 97, in <module>
    driver2 = webdriver.Chrome(service=service2)
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
    ~~~~~~~~~~~~~~~~^
        browser_name=DesiredCapabilities.CHROME["browserName"],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        keep_alive=keep_alive,
        ^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/chromium/webdriver.py", line 55, in __init__
    self.service.start()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/common/service.py", line 109, in start
    self._start_process(self._path)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/common/service.py", line 220, in _start_process
    self.process = subprocess.Popen(
                   ~~~~~~~~~~~~~~~~^
        cmd,
        ^^^^
    ...<7 lines>...
        **self.popen_kw,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/cgoldberg617/.pyenv/versions/3.13.2/lib/python3.13/subprocess.py", line 1004, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
                         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cgoldberg617/.pyenv/versions/3.13.2/lib/python3.13/subprocess.py", line 1740, in _get_handles
    c2pwrite = stdout.fileno()
ValueError: I/O operation on closed file
```

It's not a race condition, because I tried adding a long sleep before instantiating the 2nd driver and I still get the same error.

This same error occurs with all other browsers besides Chrome, so it's not browser related.

What is going ON??

### Reproducible Code

```shell
import sys
from selenium import webdriver
from selenium.webdriver.chrome.service import Service

service = Service(log_output=sys.stdout)
driver = webdriver.Chrome(service=service)
driver.quit()
service2 = Service(log_output=sys.stdout)
driver2 = webdriver.Chrome(service=service2)
```

### Debugging Logs

```logs
DEBUG:selenium.webdriver.common.selenium_manager:Selenium Manager binary found at: /home/cgoldberg617/code/selenium/py/selenium/webdriver/common/linux/selenium-manager
DEBUG:selenium.webdriver.common.selenium_manager:Executing process: /home/cgoldberg617/code/selenium/py/selenium/webdriver/common/linux/selenium-manager --browser chrome --debug --language-binding python --output json
DEBUG:selenium.webdriver.common.selenium_manager:chromedriver not found in PATH
DEBUG:selenium.webdriver.common.selenium_manager:chrome not found in PATH
DEBUG:selenium.webdriver.common.selenium_manager:chrome not found in the system
DEBUG:selenium.webdriver.common.selenium_manager:Required browser: chrome 135.0.7049.84
DEBUG:selenium.webdriver.common.selenium_manager:chrome 135.0.7049.84 already exists
DEBUG:selenium.webdriver.common.selenium_manager:chrome 135.0.7049.84 is available at /home/cgoldberg617/.cache/selenium/chrome/linux64/135.0.7049.84/chrome
DEBUG:selenium.webdriver.common.selenium_manager:Required driver: chromedriver 135.0.7049.84
DEBUG:selenium.webdriver.common.selenium_manager:chromedriver 135.0.7049.84 already in the cache
DEBUG:selenium.webdriver.common.selenium_manager:Driver path: /home/cgoldberg617/.cache/selenium/chromedriver/linux64/135.0.7049.84/chromedriver
DEBUG:selenium.webdriver.common.selenium_manager:Browser path: /home/cgoldberg617/.cache/selenium/chrome/linux64/135.0.7049.84/chrome
DEBUG:selenium.webdriver.common.service:Started executable: `/home/cgoldberg617/.cache/selenium/chromedriver/linux64/135.0.7049.84/chromedriver` in a child process with pid: 32003 using 0 to output <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
Starting ChromeDriver 135.0.7049.84 (6c019e56001911b3fd467e03bf68c435924d62f4-refs/branch-heads/7049@{#1778}) on port 53461
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully on port 53461.
DEBUG:selenium.webdriver.remote.remote_connection:POST http://localhost:53461/session {'capabilities': {'firstMatch': [{}], 'alwaysMatch': {'browserName': 'chrome', 'pageLoadStrategy': <PageLoadStrategy.normal: 'normal'>, 'browserVersion': None, 'goog:chromeOptions': {'extensions': [], 'binary': '/home/cgoldberg617/.cache/selenium/chrome/linux64/135.0.7049.84/chrome', 'args': []}}}}
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:53461
DEBUG:urllib3.connectionpool:http://localhost:53461 "POST /session HTTP/1.1" 200 0
DEBUG:selenium.webdriver.remote.remote_connection:Remote response: status=200 | data={"value":{"capabilities":{"acceptInsecureCerts":false,"browserName":"chrome","browserVersion":"135.0.7049.84","chrome":{"chromedriverVersion":"135.0.7049.84 (6c019e56001911b3fd467e03bf68c435924d62f4-refs/branch-heads/7049@{#1778})","userDataDir":"/tmp/.org.chromium.Chromium.reJVS7"},"fedcm:accounts":true,"goog:chromeOptions":{"debuggerAddress":"localhost:35293"},"networkConnectionEnabled":false,"pageLoadStrategy":"normal","platformName":"linux","proxy":{},"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify","webauthn:extension:credBlob":true,"webauthn:extension:largeBlob":true,"webauthn:extension:minPinLength":true,"webauthn:extension:prf":true,"webauthn:virtualAuthenticators":true},"sessionId":"fb374a498572f448b21fc696f5aab487"}} | headers=HTTPHeaderDict({'Content-Length': '848', 'Content-Type': 'application/json; charset=utf-8', 'cache-control': 'no-cache'})
DEBUG:selenium.webdriver.remote.remote_connection:Finished Request
DEBUG:selenium.webdriver.remote.remote_connection:DELETE http://localhost:53461/session/fb374a498572f448b21fc696f5aab487 {}
DEBUG:urllib3.connectionpool:http://localhost:53461 "DELETE /session/fb374a498572f448b21fc696f5aab487 HTTP/1.1" 200 0
DEBUG:selenium.webdriver.remote.remote_connection:Remote response: status=200 | data={"value":null} | headers=HTTPHeaderDict({'Content-Length': '14', 'Content-Type': 'application/json; charset=utf-8', 'cache-control': 'no-cache'})
DEBUG:selenium.webdriver.remote.remote_connection:Finished Request
DEBUG:selenium.webdriver.common.selenium_manager:Selenium Manager binary found at: /home/cgoldberg617/code/selenium/py/selenium/webdriver/common/linux/selenium-manager
DEBUG:selenium.webdriver.common.selenium_manager:Executing process: /home/cgoldberg617/code/selenium/py/selenium/webdriver/common/linux/selenium-manager --browser chrome --debug --language-binding python --output json
DEBUG:selenium.webdriver.common.selenium_manager:chromedriver not found in PATH
DEBUG:selenium.webdriver.common.selenium_manager:chrome not found in PATH
DEBUG:selenium.webdriver.common.selenium_manager:chrome not found in the system
DEBUG:selenium.webdriver.common.selenium_manager:Required browser: chrome 135.0.7049.84
DEBUG:selenium.webdriver.common.selenium_manager:chrome 135.0.7049.84 already exists
DEBUG:selenium.webdriver.common.selenium_manager:chrome 135.0.7049.84 is available at /home/cgoldberg617/.cache/selenium/chrome/linux64/135.0.7049.84/chrome
DEBUG:selenium.webdriver.common.selenium_manager:Required driver: chromedriver 135.0.7049.84
DEBUG:selenium.webdriver.common.selenium_manager:chromedriver 135.0.7049.84 already in the cache
DEBUG:selenium.webdriver.common.selenium_manager:Driver path: /home/cgoldberg617/.cache/selenium/chromedriver/linux64/135.0.7049.84/chromedriver
DEBUG:selenium.webdriver.common.selenium_manager:Browser path: /home/cgoldberg617/.cache/selenium/chrome/linux64/135.0.7049.84/chrome
Traceback (most recent call last):
  File "/home/cgoldberg617/code/selenium/py/sel.py", line 97, in <module>
    driver2 = webdriver.Chrome(service=service2)
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
    ~~~~~~~~~~~~~~~~^
        browser_name=DesiredCapabilities.CHROME["browserName"],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        keep_alive=keep_alive,
        ^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/chromium/webdriver.py", line 55, in __init__
    self.service.start()
    ~~~~~~~~~~~~~~~~~~^^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/common/service.py", line 109, in start
    self._start_process(self._path)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/cgoldberg617/code/selenium/py/selenium/webdriver/common/service.py", line 220, in _start_process
    self.process = subprocess.Popen(
                   ~~~~~~~~~~~~~~~~^
        cmd,
        ^^^^
    ...<7 lines>...
        **self.popen_kw,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/cgoldberg617/.pyenv/versions/3.13.2/lib/python3.13/subprocess.py", line 1004, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
                         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cgoldberg617/.pyenv/versions/3.13.2/lib/python3.13/subprocess.py", line 1740, in _get_handles
    c2pwrite = stdout.fileno()
ValueError: I/O operation on closed file
```

### PR DESCRIPTION
**Pull request by jit3pam**
### 🔗 Related Issues

Fixes #15629 
### 💥 What does this PR do?

This change fixes an issue where the Service.stop() method closes the stream provided via the log_output parameter even when the stream was not created by Selenium. 
I could replicate the errors mentioned in #15629.

When users pass streams such as sys.stdout, the current implementation closes the global stream when the service stops. It causes error in new WebDriver instances that try to use the same stream.

The fix introduces tracking of the ownership of the log_output stream. 
With this fix Selenium now closes the stream only if it was opened internally (for example when a file path is provided). Streams provided by the user, such as sys.stdout or other file-like objects, will not be closed by the stop method defined in Service class.

I have also added a test to verify that multiple WebDriver instances can be created sequentially using Service(log_output=sys.stdout) without failure.

### 🔄 Types of changes

- Bug fix (backwards compatible)
- New feature (non-breaking change which adds functionality *and tests!*)

### CODE REVIEW

File: py/test/selenium/webdriver/common/test_service_logging.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import sys
+import pytest
+
+from selenium import webdriver
+from selenium.webdriver.chrome.service import Service
+
+@pytest.mark.chrome
+def test_service_allows_reusing_stdout_for_logging():
```
[the remark this answers is not in the document]
Comment (jit3pam) [reply py/test/selenium/webdriver/common/test_service_logging.py]: removed the `test_service_logging.py` file. I have addd the test in the existing `chrome_service_testing.py`

File: py/test/selenium/webdriver/common/test_service_logging.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    try:
+        service1 = Service(log_output=sys.stdout)
+        browser1 = webdriver.Chrome(service=service1) # lazy_import
+        assert browser1.session_id is not None
+        browser1.quit()
+        browser1 = None
+
+        service2 = Service(log_output=sys.stdout)
+        browser2 = webdriver.Chrome(service=service2) # lazy_import
+        assert browser2.session_id is not None
```
[the remark this answers is not in the document]
Comment (jit3pam) [reply py/test/selenium/webdriver/common/test_service_logging.py]: removed the test_service_logging.py file. I have addd the test in the existing chrome_service_testing.py

**Review by cgoldberg — CHANGES_REQUESTED** [review]
Thanks... looks good... just a few minor things:

1. Can you remove the blank lines in your test, so it looks like other tests in that file?
2. The CI Lint job is going to fail on this due to some minor formatting issues. You can fix them by running `./scripts/format.sh` (if you have bazel setup)... or `tox -c ./py/tox.ini -e linting` (`pip install tox` if you need it)
3. Since this affects the base `Service` class, we should probably test it on other browsers. Can you add a similar test to:
    - `./py/selenium/webdriver/edge/edge_service_tests.py`
    - `./py/selenium/webdriver/firefox/firefox_service_tests.py`
    - `./py/selenium/webdriver/safari/safari_service_tests.py`

**Review by jit3pam — COMMENTED** [review]
@cgoldberg I have added more tests and also followed linting suggestions using tox.
Please review and let me know if more changes are required. 
Thank you in advance.

File: py/test/selenium/webdriver/chrome/chrome_service_tests.py [hunk 3]
```diff
@@ -123,6 +123,17 @@ def test_driver_is_stopped_if_browser_cant_start(clean_driver, clean_options, dr
     assert service.process.poll() is not None
 
 
+def test_service_allows_reusing_stdout_for_logging(clean_driver, clean_options, driver_executable) -> None:
+    service1 = Service(executable_path=driver_executable, log_output=sys.stdout)
+    driver1 = clean_driver(service=service1, options=clean_options)
+    assert driver1.session_id is not None
+    driver1.quit()
+    driver1 = None
```
Comment (cgoldberg) [inline py/test/selenium/webdriver/chrome/chrome_service_tests.py]: I don't think you need this line

File: py/test/selenium/webdriver/chrome/chrome_service_tests.py [hunk 4]
```diff
@@ -123,6 +123,17 @@ def test_driver_is_stopped_if_browser_cant_start(clean_driver, clean_options, dr
...
+    service2 = Service(executable_path=driver_executable, log_output=sys.stdout)
+    driver2 = clean_driver(service=service2, options=clean_options)
+    assert driver2.session_id is not None
+
```
Comment (cgoldberg) [inline py/test/selenium/webdriver/chrome/chrome_service_tests.py:143]: forgot `driver2.quit()`?

**Review by cgoldberg — CHANGES_REQUESTED** [review]
few more tiny changes...

Also, I think it was better the way you were quitting the browser in a `finally` block in your previous version... Can you re-add that?

File: py/test/selenium/webdriver/chrome/chrome_service_tests.py (hunk 4 above)
Comment (jit3pam) [reply py/test/selenium/webdriver/chrome/chrome_service_tests.py:143]: I have added `browser2.quit()`.

File: py/test/selenium/webdriver/chrome/chrome_service_tests.py (hunk 3 above)
Comment (jit3pam) [reply py/test/selenium/webdriver/chrome/chrome_service_tests.py]: I have removed that line in which browser1 variable was assigned None.

**Comment by jit3pam** [comment]
> few more tiny changes...
> 
> Also, I think it was better the way you were quitting the browser in a `finally` block in your previous version... Can you re-add that?

I have updated the test methods with `finally` block. Infact I have restored the previous version. 
I have also added the same test method for firefox, safari, and edge.

**Review by cgoldberg — APPROVED** [review]
@jit3pam LGTM 👍 

Thanks for being responsive with your updates. I'll merge this once CI runs.

**Comment by cgoldberg** [comment]
The failure in CI is unrelated.. all the Python tests ran fine. Merging now.
